Inactive shard flush should wait for ongoing one

kingherc opened the pull request:
org.elasticsearch.indices.flush.FlushIT#testFlushOnInactive would
sometimes fail in the following case:
* SHARD_MEMORY_INTERVAL_TIME_SETTING is set very low, e.g., 10ms
* The regularly scheduled multiple flushes proceed to
  org.elasticsearch.index.shard.IndexShard#flushOnIdle
* There, the first flush will handle e.g., the first document
  that was indexed. The second flush will arrive shortly after,
  before the first flush finishes.
* The second flush will find that wasActive = true (due to the
  indexing of the remaining documents), and will set it to false.
* However, the second flush will not be executed because
  waitIfOngoing = false, and there is the ongoing first flush.
* No other flush is scheduled (since any next regularly scheduled
  flush will find wasActive = false), which creates the problem.

Fixes #87888

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

kingherc commented:
Hi @henningandersen and @DaveCTurner . After investigating the test failure, we (with the help of @fcofdez ) figured out the situation, which is presented in the PR description. It can happen in some rare situations when the SHARD_MEMORY_INTERVAL_TIME_SETTING is set too low and/or if a flush takes too long.

There are 3 solutions we discussed so far:
1. The current PR's solution, to change the waitIfOngoing to true. Drawback: one of the flush threadpool's threads may wait on the flushLock. But for normal values of SHARD_MEMORY_INTERVAL_TIME_SETTING (e.g. the default is 5 seconds), this drawback would appear in some very weirdly long-running flush that was being executed already. This is why the PR has this solution.
2. To avoid the drawback of the first solution, we could "chain" a second flush request to the first ongoing request. And make the first request, when finished, execute any chained requests. That way, we do not have a thread waiting on the flushLock. Drawback: complexity vs the other approaches.
3. We remove the wasActive smartness. Drawback: every SHARD_MEMORY_INTERVAL_TIME_SETTING ms, there will be a possible no-op flush request scheduled.

Feel free to tell us your opinion on the selected approach and any other thoughts you may have.

henningandersen commented:
It does look like the old synced-flush would use `waitIfOngoing=true`, see [here](https://github.com/elastic/elasticsearch/blob/bcfbf000748b538884087623113682165abcb92d/server/src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java#L589). However, I wonder if we could let `flush` return a boolean if it attempted the flush or not - and if not, set `active` back to true?

A more straightforward solution could also be to move the setting of `active` to be after indexing is done rather than before? This also affects `scheduledRefresh`, but I think it is fine to not consider an indexing request that is ongoing active there too. I think I'd prefer that option, unless I am overseeing something.

fcofdez commented:
> A more straightforward solution could also be to move the setting of active to be after indexing is done rather than before? This also affects scheduledRefresh, but I think it is fine to not consider an indexing request that is ongoing active there too. I think I'd prefer that option, unless I am overseeing something.

I'm not sure if this solves the issue? Theoretically we could still miss the last flush if the first flush is still running after the remaining documents have been indexed. But maybe I'm missing something here.

henningandersen commented:
> I'm not sure if this solves the issue? Theoretically we could still miss the last flush if the first flush is still running after the remaining documents have been indexed. But maybe I'm missing something here.

You might be right :slightly_smiling_face: . The idea would be that by marking active after the indexing has occurred, the next round of `flushOnIdle` would know that it needs to flush?

kingherc commented:
Hi! Thanks for the awesome conversation. I think I agree with @fcofdez . If we follow that approach @henningandersen , I so see a rare situation where:
* There's an ongoing flush of docs [1,N]
* A new doc N+1 is indexed and sets active = true
* The next flushOnIdle() is called, which sets active = false, and tries another flush request, which fails because of the ongoing flush.
* The ongoing flush finishes, but has missed N+1 document.
* Any next scheduled flushOnIdle() will not try flush requests since active = false.
Right?

> However, I wonder if we could let flush return a boolean if it attempted the flush or not - and if not, set active back to true?

I think this is also a nice solution. I can try that if you agree.

fcofdez commented:
Just to add more context here, this only affects in cases where we stop indexing after the latest flush is skipped (a rare edge case)

> I think this is also a nice solution. I can try that if you agree.

👍 let's try that.

kingherc committed:
Inactive shard flush should wait for ongoing one

org.elasticsearch.indices.flush.FlushIT#testFlushOnInactive would
sometimes fail in the following case:
* SHARD_MEMORY_INTERVAL_TIME_SETTING is set very low, e.g., 10ms
* The regularly scheduled multiple flushes proceed to
  org.elasticsearch.index.shard.IndexShard#flushOnIdle
* There, the first flush will handle e.g., the first document
  that was indexed. The second flush will arrive shortly after,
  before the first flush finishes.
* The second flush will find that wasActive = true (due to the
  indexing of the remaining documents), and will set it to false.
* However, the second flush will not be executed because
  waitIfOngoing = false, and there is the ongoing first flush.
* No other flush is scheduled (since any next regularly scheduled
  flush will find wasActive = false), which creates the problem.

Solution: if a flush request does not happen, revert active flag,
so that a next flush request can happen.

Fixes #87888

henningandersen commented:
Makes sense, the `active` flag would need to be cleared differently too (inside flush) if were to pursue the solution where we set active after indexing.

kingherc commented:
Hi @fcofdez , @henningandersen . Thanks for the conversation. The method where if flush does not wait for the ongoing flush, it returns false, and we set the active flag back to true, works. I did that -- feel free to review the PR.

fcofdez commented:
It looks like there are some related test failures, additionally we usually avoid force-pushing since Github gets confused and it hides/removes some review comments.

kingherc committed:
Fix test

kingherc commented:
> It looks like there are some related test failures, additionally we usually avoid force-pushing since Github gets confused and it hides/removes some review comments.

Fixed the test.

Oh, about the force pushes, I will avoid from now on. Either way the commits get squashed when merging the PR, so indeed I do not see a reason now why I did it :) 
Did you miss adding a comment? if yes, please mention it again please.

henningandersen commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1355,8 +1356,11 @@ public void flush(FlushRequest request) {
-        getEngine().flush(force, waitIfOngoing);
-        flushMetric.inc(System.nanoTime() - time);
+        boolean flushHappened = getEngine().flush(force, waitIfOngoing);
+        if (flushHappened) {
```
I am in doubt about this. The [documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-nodes-stats.html#cluster-nodes-stats-api-response-body) says this is "Number of flush operations". So far it has been the same as API calls, now it is only those that do anything. It is a subtle change and hardly breaking.

But this change puts the flush-metric between API calls and actual flushes. It is not actual flushes, because if there is nothing to flush, we will still increase it here.

So I'd prefer to inc always here and then we can always make it more accurate.

kingherc replied:
Done

henningandersen commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2188,8 +2192,12 @@ public void onFailure(Exception e) {
-                        flush(new FlushRequest().waitIfOngoing(false).force(false));
-                        periodicFlushMetric.inc();
+                        if (flush(new FlushRequest().waitIfOngoing(false).force(false))) {
+                            periodicFlushMetric.inc();
```
I think the same argument as above holds here too.

henningandersen reviewed (commented):
This directions looks good to me. I have a few comments that I'd like to see addressed though.

fcofdez commented on server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java:
```diff
@@ -437,8 +437,8 @@ public boolean shouldPeriodicallyFlush() {
-    public void flush(boolean force, boolean waitIfOngoing) throws EngineException {
-        // noop
+    public boolean flush(boolean force, boolean waitIfOngoing) throws EngineException {
+        return true; // noop
```
I think this should return false? This implementation won't flush.

henningandersen replied:
But that is not really the meaning of the return value as it stands in `InternalEngine`. `InternalEngine` may also return true without flushing. What the return value currently mean is

false: if `waitIfOngoing` detected a collision
true: otherwise.

Here true seems most correct since no collision was detected.

fcofdez replied:
Yes, I was wrong, please dismiss my comment.

fcofdez reviewed (commented):
I left a small comment, the direction looks good.

kingherc committed:
New test for concurrent flushes

And fix some PR review feedback

kingherc commented on server/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -1015,8 +1015,9 @@ public boolean refreshNeeded() {
      * @param force         if <code>true</code> a lucene commit is executed even if no changes need to be committed.
      * @param waitIfOngoing if <code>true</code> this call will block until all currently running flushes have finished.
      *                      Otherwise this call will return without blocking.
+     * @return <code>true</code> if the flush happened, else <code>false</code> (e.g., if it did not wait for an ongoing request)
```
Done

henningandersen commented on server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -1908,7 +1908,8 @@ public void flush(boolean force, boolean waitIfOngoing) throws EngineException {
                 // if we can't get the lock right away we block if needed otherwise barf
                 if (waitIfOngoing == false) {
-                    return;
+                    logger.trace("returning as there is an in-flight flush that we do not need to wait for");
```
nit: I'd like the log message to be less about the "return" statement and more about the situation:

```suggestion
                    logger.trace("detected an in-flight flush, not blocking to wait for it's completion");
```

henningandersen commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1343,8 +1343,9 @@ public BulkStats bulkStats() {
      * Executes the given flush request against the engine.
      *
      * @param request the flush request
+     * @return <code>false</code> if the flush did not wait for an ongoing request and returned, else <code>true</code>
```
```suggestion
     * @return <code>false</code> if waitIfOngoing==true and an ongoing request is detected, else <code>true</code>. If false is returned, no flush happened.
```

henningandersen commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2188,7 +2190,10 @@ public void onFailure(Exception e) {
                     protected void doRun() {
-                        flush(new FlushRequest().waitIfOngoing(false).force(false));
+                        if (flush(new FlushRequest().waitIfOngoing(false).force(false)) == false) {
+                            // In case the flush did not happen, revert active flag so that a next flushOnIdle request can happen (#87888)
```
```suggestion
                            // In case an ongoing flush was detected, revert active flag so that a next flushOnIdle request will retry (#87888)
```

henningandersen commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -3897,6 +3898,63 @@ public void testFlushOnIdle() throws Exception {
+            );
+            assertBusy(mockLogAppender::assertAllExpectationsMatched);
+
+            // While the previous flushOnIdle request is happening, immediately issue a second flush request with waitIfOngoing=false
```
I think we need to block the original flush in order to be sure this happens?

kingherc replied:
Hi @henningandersen , indeed, that would be best. For the moment, I've indexed a large-enough amount to ensure the flush is still ongoing. Would you have a recommendation about how to inject a block of the original flush just for the test (rather than in reality)?

kingherc replied:
After some tips from @fcofdez , I found a nice way to do it. Feel free to review :)

henningandersen reviewed (commented):
I think there is a problem with the new test, otherwise this looks good.

kingherc committed:
Apply suggestions from code review

Co-authored-by: Henning Andersen <33268011+henningandersen@users.noreply.github.com>

kingherc committed:
Fix test and comments

kingherc committed:
Merge remote-tracking branch 'origin/main' into test-failure/87888-flush-on-inactive

henningandersen reviewed (approved):
LGTM.

fcofdez reviewed (approved):
LGTM 👍

kingherc committed:
Test active flag as well

Fix some javadoc

kingherc commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -3897,6 +3899,78 @@ public void testFlushOnIdle() throws Exception {
+                )
+            );
+            assertBusy(mockLogAppender::assertAllExpectationsMatched);
+
```
Hi @henningandersen , could not do it immediately because I called flush() instead of a flushOnIdle request. Modified the test slightly to also do another flushOnIdle request to test that the active flag turns true. Thanks for the suggestion!

kingherc commented:
@elasticmachine run elasticsearch-ci/part-1 please

henningandersen commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -3897,6 +3899,89 @@ public void testFlushOnIdle() throws Exception {
+            );
+            assertBusy(mockLogAppender::assertAllExpectationsMatched);
+
+            // The second flush (that did not happen) should have returned false and turned the active flag to true
```
I am not sure I understand this comment, is `active` not only set/reset in `flushOnIdle` and not in `flush`?

kingherc replied:
Hi @henningandersen , `active` is only toggled during the `flushOnIdle()` function (which later calls `flush()`) and **not** during `flush()` itself.

henningandersen replied:
Perhaps change comment to be about `flushOnIdle`? The second `flush` is easy to mistake for the flush a few lines up...

kingherc replied:
Oh I see, sure! I did that and improved a bit the comments.

kingherc committed:
Improve test comments
